src/liboslcomp/typecheck.cpp

```diff
@@ -946,9 +946,17 @@ ASTfunction_call::mark_optional_output (int firstopt, const char **tags)
        ASTNode *s = argvec[a].get();
        bool isoutput = false;
        // compare against output tags
-       if (s->typespec().is_string() && s->nodetype() == ASTNode::literal_node) {
-           for (const char **tag = tags; *tag && !isoutput; ++tag)
-               isoutput = isoutput || mark_all || (! strcmp (((ASTliteral *)s)->strval(), *tag));
+       if (s->typespec().is_string()) {
+           if (s->nodetype() == ASTNode::literal_node) {
+               // If the token is a string literal, see if it's one of the
+               // ones designated as an output slot.
+               for (const char **tag = tags; *tag && !isoutput; ++tag)
+                   isoutput = isoutput || mark_all || (! strcmp (((ASTliteral *)s)->strval(), *tag));
+           } else {
+               // If the token is not a literal, we don't know what it'll
+               // be at runtime, so mark it conservatively as possible output.
+               isoutput = true;
```

Comment:
Is `pointcloud_search` the only function affected? Or were other ones buggy as well?

Are there any functions that take optional args but don't have any optional outputs?

Reply:
They were all buggy in the same way. It could easily(*) have affected a texture() (and environment(), etc.) call if the "alpha" or "errormessage" optional parameter used something other than a string literal as its token.

But "easily" should be taken with a grain of salt -- it's a bit of a weird edge case for those not to be string literals. It's possible, but very unusual. So I'm not tempted to fret over the possibility that there might be values conservatively marked as "potentially written" that in fact are not.

The mark_optional_output function in question is only called for texture (family) and pointcloud_get/search, both of which have this property of potentially writeable variable references.